lib/active_record/connection_adapters/sqlserver/database_statements.rb

```diff
@@ -206,6 +206,35 @@ def build_sql_for_merge_insert(insert:, insert_all:, columns_with_uniqueness_con
           sql
         end
 
+        # For `nil` identity columns we need to ensure that the values do not match so that they are all inserted.
+        # Method is a combination of `ActiveRecord::InsertAll#values_list` and `ActiveRecord::ConnectionAdapters::SQLServer::DatabaseStatements#default_insert_value`.
+        def merge_insert_values_list(insert:, insert_all:)
+          connection = insert.send(:connection)
+          identity_index = 0
+
+          types = insert.send(:extract_types_from_columns_on, insert.model.table_name, keys: insert.keys_including_timestamps)
+
+          values_list = insert_all.map_key_with_value do |key, value|
+            if Arel::Nodes::SqlLiteral === value
+              value
+            elsif insert.primary_keys.include?(key) && value.nil?
+              column = insert.send(:column_from_key, key)
+
+              if column.is_identity?
+                identity_index += 1
+                table_name = quote(quote_table_name(column.table_name))
+                Arel.sql("IDENT_CURRENT(#{table_name}) + (IDENT_INCR(#{table_name}) * #{identity_index})")
```

Comment:
shouldn't this be enough?

```suggestion
                Arel.sql("IDENT_CURRENT(#{table_name}) + (IDENT_INCR(#{table_name}) + #{identity_index})")
```

but I am not sure. I assume MSSQL uses the identifier generated here and does not overwrite them, correct? if it does overwrite it, then ignore this comment. if it does not overwrite it, then I would prefer the `+`, as it would generate much smaller values.

Reply:
The values given by `"IDENT_CURRENT(#{table_name}) + (IDENT_INCR(#{table_name}) + #{identity_index})"` are the IDs used when the record is created. I used multiplication as `IDENT_INCR` is the step value (https://learn.microsoft.com/en-us/sql/t-sql/functions/ident-incr-transact-sql?view=sql-server-ver16) that should be used for generating the ID. 

So if `IDENT_CURRENT('books)` is 100 and `IDENT_INCR('books')` is 2 then the next IDs generated should be 102, 104, 106, etc. To generate this sequence we need to use multiplication. If we used addition then the IDs generated would be 102, 103, 104, etc.